src/Functions/array/arrayShuffle.cpp

```diff
@@ -0,0 +1,283 @@
+
+    IColumn & res_data = res.getData();
+
+    const ColumnNullable * src_nullable_col = typeid_cast<const ColumnNullable *>(&src_data);
+    ColumnNullable * res_nullable_col = typeid_cast<ColumnNullable *>(&res_data);
+
+    const IColumn * src_inner_col = src_nullable_col ? &src_nullable_col->getNestedColumn() : &src_data;
+    IColumn * res_inner_col = res_nullable_col ? &res_nullable_col->getNestedColumn() : &res_data;
+
+    const auto seed = [&]() -> uint64_t
+    {
+        if (arguments.size() == 1)
+            return randomSeed();
+        const auto * val = arguments[1].column.get();
+        return val->getUInt(0);
+    }();
+    pcg64_fast rng(seed);
+
+    false // NOLINT
+        || executeNumber<UInt8>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<UInt16>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<UInt32>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<UInt64>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Int8>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Int16>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Int32>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Int64>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Float32>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeNumber<Float64>(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeString(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeFixedString(*src_inner_col, offsets, *res_inner_col, rng)
+        || executeGeneric(*src_inner_col, offsets, *res_inner_col, rng);
+
+    if (src_nullable_col)
+    {
+        rng.seed(seed);
+        if (!executeNumber<UInt8>(src_nullable_col->getNullMapColumn(), offsets, res_nullable_col->getNullMapColumn(), rng))
+            throw Exception(
+                "Illegal column " + src_nullable_col->getNullMapColumn().getName() + " of null map of the first argument of function "
+                    + getName(),
+                ErrorCodes::ILLEGAL_COLUMN);
+    }
+
+    return res_ptr;
+}
+
+bool FunctionArrayShuffle::executeGeneric(const IColumn & src_data, const ColumnArray::Offsets & src_offsets, IColumn & res_data, pcg64_fast & rng)
+{
+    size_t size = src_offsets.size();
+    res_data.reserve(size);
+
+    IColumn::Permutation permutation;
+    ColumnArray::Offset prev_off = 0;
+    for (auto off: src_offsets)
+    {
+        size_t count = off - prev_off;
+
+        permutation.resize(count);
+        for (size_t idx = 0; idx < count; ++idx)
+            permutation[idx] = idx;
+        std::shuffle(std::begin(permutation), std::end(permutation), rng);
+        for (size_t unshuffled_idx = 0; unshuffled_idx != count; ++unshuffled_idx)
+        {
+            auto shuffled_idx = permutation[unshuffled_idx];
+            res_data.insertFrom(src_data, shuffled_idx);
```

Comment:
Can we use the IColumn::permute method?

Reply:
So, we will create a big permutation for all the arrays' data, but shuffle this permutation only inside each array, and then use it to permute the data.

This method might appear to be as efficient as specializations for various types.

Reply:
Should all types be handled in a generic way (`IColumn::permute`) - as in the latest commit - or choose some mid-ground -> e.g. specialised numbers + rest generic?

Reply:
The only difference would be that for numbers we don't need to create the permutations array. Instead we `memcpy` the data and shuffle directly in the result (+ optionally the null column). 
Previous logic can be restored by reverting: [FIXUP: Simplify logic by using permute function](https://github.com/ClickHouse/ClickHouse/pull/45271/commits/cf1c5e4ac0b506789d7744a34acc24d3d06bfe6f)